packages/astro/src/vite-plugin-astro-server/index.ts

```diff
@@ -79,37 +89,67 @@ async function handleRequest(
 	const origin = `${viteServer.config.server.https ? 'https' : 'http'}://${req.headers.host}`;
 	const pathname = decodeURI(new URL(origin + req.url).pathname);
 	const rootRelativeUrl = pathname.substring(devRoot.length - 1);
+
 	try {
 		if (!pathname.startsWith(devRoot)) {
-			info(logging, 'astro', msg.req({ url: pathname, statusCode: 404 }));
+			log404(logging, pathname);
 			return handle404Response(origin, config, req, res);
 		}
 		// Attempt to match the URL to a valid page route.
 		// If that fails, switch the response to a 404 response.
 		let route = matchRoute(rootRelativeUrl, manifest);
 		const statusCode = route ? 200 : 404;
-		// If no match found, lookup a custom 404 page to render, if one exists.
+
 		if (!route) {
-			const relPages = config.pages.href.replace(config.projectRoot.href, '');
-			route = manifest.routes.find((r) => r.component === relPages + '404.astro');
+			log404(logging, pathname);
+			const custom404 = getCustom404Route(config, manifest);
+			if (custom404) {
+				route = custom404;
+			} else {
+				return handle404Response(origin, config, req, res);
+			}
 		}
-		// If still no match is found, respond with a generic 404 page.
-		if (!route) {
-			info(logging, 'astro', msg.req({ url: pathname, statusCode: 404 }));
-			handle404Response(origin, config, req, res);
-			return;
+
+		// Note: may re-assign filePath to custom 404 on error
+		let filePath = new URL(`./${route.component}`, config.projectRoot);
+		// Note: may re-assign preloaded component to custom 404 on error
+		let preloadedComponent = await preload({ astroConfig: config, filePath, viteServer });
+		try {
+			const [, mod] = preloadedComponent;
+			// attempt to get static paths
+			// if this fails, we have a bad URL match!
+			await getParamsAndProps({
+				mod,
+				route,
+				routeCache,
+				pathname,
+				logging,
+			});
+		} catch (_err: any) {
+			if (_err instanceof Error) {
```

Comment:
- what if something other than an `Error` object gets thrown? It looks like it passes through here
- what if an error other than a 404 gets thrown? We've had trouble in the past incorrectly swallowing errors that happen in our dependencies, due to a bug in our code, etc.

Reply:
Fixed! [Introduced a semantic enum](https://github.com/withastro/astro/pull/2743/commits/681ebcdce1b6c0fe1f6be3df863b65c24869a23d) we can match on as a return value. Should prevent unexpected errors in the future.